### ISSUE TITLE
Npm always erroneously installing devDependencies of dependency modules

### ISSUE THREAD

**Issue by fresheneesz** [npm/npm#11039]
On npm v3.3.12 (the latest that comes with node from nvm), I have this package.json

```
{
    "name": "test",
    "version": "0.1.0",
    "description": "",
    "dependencies": {
        "proto":"1.0.19",
        "gem":"2.0.2"
    }
}
```

When doing an `npm install` in the directory its in, I'm getting a devDependency of trimArguments (which is a regular dependency of `gem`) installed at `node_modules/trimArguments/node_modules/deadunit`. This is causing me a problem because when I then do an `npm shrinkwrap` it complains about the devDependency being "extraneous" and refuses to create the shrinkwrap.

**Comment by iarna** [comment]
TLDR: This is definitely a bug. You can work around this by changing your deps with shrinkwraps to use `"*"` instead of `""`.

---

Trying to reproduce this w/  `npm@3.5.2` (`npm@latest`) and I don't get quite the same results. I get this tree:

```
├─┬ gem@2.0.2
│ ├── hashmap@2.0.4
│ ├─┬ observe@1.3.5
│ │ └── proto@1.0.11
│ └── trimArguments@1.0.1
└── proto@1.0.19
```

`deadunit` is not installed (and shouldn't be– it's a dependency's dev dep and those should never be installed)

However, I am seeing this:

```
npm ERR! invalid: have proto@1.0.11 (expected: ) /Users/rebecca/code/npmtest/11039/node_modules/observe/node_modules/proto
```

And this is a new bug: `observe` has a dependency on `proto` and doesn't specify ANY version. 

Previously this was understood to mean `*`, and when validating installs ranges were just matched against the version on disk, so this worked fine. Recently not specifying a version was changed to mean `latest` and THAT we can't ordinarily meaningfully match in an ongoing way since you can change what `latest` means at any time.

For tags like `latest`, we detect them by looking inside the installed module at metadata we write there at install time. If it matches then we count it as valid. This works fine ordinarily– you run `npm install` and it stores metadata based on what your `package.json` asked for…

Except, in the case of `npm-shrinkwrap.json` we don't use the specifier from your `package.json`, we use the tarball if one was included, failing that we do `package@version` if a version is there, failing THAT, we use the from field. Only the last one gets an install npm considers valid, and even then that will fail if the spec was `""` for unrelated reasons.

You absolutely should be able to shrinkwrap things that depend on tags, and while depending on `""` is unusual, it should work too.

**Comment by fresheneesz** [comment]
Thanks for the info.

"`Recently not specifying a version was changed to mean latest`"

That's a pretty huge breaking change isn't it? I'm not sure that was my issue here, but couldn't that cause a lot of people huge headaches?

**Comment by iarna** [comment]
That was the headliner for npm@3.4.1: `https://github.com/npm/npm/releases/tag/v3.4.1`

We made the change because having it default to `*` made it under the best of circumstances behave in unexpected ways, and in the worst just break, if the latest version you had was a prerelease, because `*` doesn't match prereleases.

I don't believe you would have seen any break if we didn't ALSO have this shrinkwrap bug that was causing modules with a `dist-tag` as a version to produce trees npm considers invalid when installed via shrinkwrap file.
